server/src-test/Hasura/Server/AuthSpec.hs

```diff
@@ -0,0 +1,408 @@
+              `shouldReturn` Left AccessDenied
+            getUserInfoWithExpTime claim [("Authorization", "IGNORED"), (userRoleHeader, "admin")] mode
+              `shouldReturn` Left AccessDenied
+
+          -- A corner case, but the behavior seems desirable:
+          it "always rejects when token has empty allowedRolesClaim" $ do
+            let claim = unObject [ allowedRolesClaim .= ([] :: [Text]), defaultRoleClaim .= ("user" :: Text) ]
+            getUserInfoWithExpTime claim [("Authorization", "IGNORED"), (userRoleHeader, "admin")] mode
+              `shouldReturn` Left AccessDenied
+            getUserInfoWithExpTime claim [("Authorization", "IGNORED"), (userRoleHeader, "user")] mode
+              `shouldReturn` Left AccessDenied
+            getUserInfoWithExpTime claim [("Authorization", "IGNORED")] mode
+              `shouldReturn` Left AccessDenied
+
+          it "rejects when token doesn't have proper allowedRolesClaim and defaultRoleClaim" $ do
+            let claim0 = unObject [ allowedRolesClaim .= (["editor","user", "mod"] :: [Text]) ]
+                claim1 = unObject [ defaultRoleClaim .= ("user" :: Text) ]
+                claim2 = unObject []
+            for_ [claim0, claim1, claim2] $ \claim ->
+              getUserInfoWithExpTime claim [("Authorization", "IGNORED")] mode
+                `shouldReturn` Left JWTRoleClaimMissing
+
+    -- (*) FIXME NOTE (re above):
+    --
+    -- Ideally we should always return AccessDenied if the role we would
+    -- otherwise have returned does not match the requested role (from the
+    -- 'userRoleHeader').
+    --
+    -- This would harden a bit against bugs, makes the spec simpler, but
+    -- especially is better UX since in the current behavior the user can't be
+    -- sure which role their query is operating as (and in the worst case we
+    -- might e.g. delete rows the user didn't intend)
+    --
+    -- But this is a breaking change we need to think a little more about;
+    -- users might be relying on the behavior, e.g. just hardcoding a dev role
+    -- into clients.
+
+
+-- Some very basic unit tests of AuthMode construction and error modes
+setupAuthModeTests :: Spec
+setupAuthModeTests = describe "setupAuthMode" $ do
+  let secret = hashAdminSecret "secret"
+      unauthRole = mkRoleNameE "anon"
+
+  -- These are all various error cases, except for the AMNoAuth mode:
+  it "with no admin secret provided" $ do
+    setupAuthMode' Nothing       Nothing             Nothing              Nothing
+      `shouldReturn` (Right AMNoAuth)
+    -- We insist on an admin secret in order to use webhook or JWT auth:
+    setupAuthMode' Nothing       Nothing             (Just fakeJWTConfig) Nothing
+      `shouldReturn` Left ()
+    setupAuthMode' Nothing       (Just fakeAuthHook) Nothing              Nothing
+      `shouldReturn` Left ()
+    -- ...and we can't have both:
```

Comment:
I just want to add here that I think the comments here are really useful, especially since it's often difficult to see if the output should be `Left` or `Right`. I can't say I'm convinced that all of these are right, but that is due to my brain not being able to parse such a long list of outputs. But at least this gives a clear reference that we can check against, together with motivations for each test.